platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ExportMigrationConfigurationAdminContext.java

```diff
@@ -0,0 +1,208 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.configuration.admin;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.lang.Validate;
+import org.apache.felix.fileinstall.internal.DirectoryWatcher;
+import org.codice.ddf.configuration.persistence.PersistenceStrategy;
+import org.codice.ddf.migration.ExportMigrationContext;
+import org.codice.ddf.migration.ExportMigrationContextProxy;
+import org.codice.ddf.migration.ExportMigrationEntry;
+import org.codice.ddf.migration.MigrationEntry;
+import org.codice.ddf.migration.MigrationWarning;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class extends on the {@link ExportMigrationContext} interface to pre-create entries for
```

Comment:
❓ This is a design question that came to mind as I was reviewing the other classes.

The design is built around the idea of a Context that can be used to get configuration entries, which can be exported and imported back.

The default implementation deals with files on the file system and relies on the migratables themselves to get entries for the files they need to process. In other words, the Context _doesn't know anything_ about the list of files migratables have to deal with.

However, for `Configuration` objects, the approach seems different. The Context _knows_ what's available and the `ConfigurationAdminMigratable` is almost just an empty shell that delegates all the work to the Context.

This impacts the design in a few different ways:
* The logic and approach used by the `ConfigurationAdminMigratable` is different than the other migratables, which makes the solution more difficult to understand since the responsibilities are shifted.
* It opened the door to the creation of many `Proxy` classes in the API bundle, which adds complexity to the design and maintenance costs. Keeping the same separation of concerns would most likely make those `Proxy` classes unnecessary.
* It forces us to keep track of the `ImportMigrationConfigurationAdminContext`'s validity using the `isValid` flag and extra logic, which should be easier to handle in the `ConfigurationAdminMigratable` class itself.
* If the code cannot be written without having to reimplement new Context and Entry classes to deal with `Configuration` objects, it feels like the API and default implementation aren't as generic as they should be.

Thoughts?

Reply:
I opted for this approach here because what needed to be processed at import time was not just what was exported. In some case, additional config objects might need to be deleted. I needed one entity above the entries that could do that and I opted for the context. 

From the perspective of the migratable, there is no difference than in all the others. It asks for all migratable entries (might be getting more than was exported) and simply processed them one by one. 

It also allows for the context to better keep track of which one needs updating, creation, or deletion.